Refactoring

jorisdral opened the pull request:
Changes include:
* Update GHA workflow file
* Running CI for GHC versions up to `9.8.2`
* Delete the `Setup.hs` file
* Update cabal file:
  * Constrain base for the GHC versions we're testing with in CI
  * Constrain the version of lmdb to be `0.9.x`
* Extract the FFI bindings and types from the `Raw` module and put them in a separate `FFI` module in a private sub-library
* Changes to the FFI bindings: 
  * Use `capi` instead of `ccall`, because `capi` is recommended
    * One exception: `mdb_env_get_path` is still `ccall`, because with `capi` it would require the use of `ConstPtr`, which does not exist in GHC's before `9.6`
  * Import values of flags/enums using `capi` instead of using `#const`
  * Use newtypes for `typedef`s 
  * Create empty types for opaque structs
  * Favour qualified imports (`FFI.`) for FFI definitions instead of prefixing by `_`
* Changes to higher-level bindings in `Raw.hs`:
   * Rework compiling and de-compiling of flags to not use arrays
   * Use `FFI.` qualified imports
   * Fix compilation: change some types, do some wrapping and unwrapping of newtypes
 * Add a simple test for both the low-level `FFI` bindings and the higher-level `Raw` bindings.
 * Change the source file folder to `src` from `hsrc_lib`

jorisdral commented:
https://github.com/input-output-hk/haskell-lmdb/actions/runs/8469238595/job/23204871467 errors because it says `async` is an unused package, but that is not the case. I'm not sure what's going on here

jorisdral commented on hsrc_lib/Database/LMDB/FFI.hsc:
```diff
@@ -0,0 +1,293 @@
+  Cursor get operations
+-------------------------------------------------------------------------------}
+
+-- TODO: can I get rid of the newtype?
```
```suggestion
```

jorisdral commented on hsrc_lib/Database/LMDB/FFI.hsc:
```diff
@@ -0,0 +1,293 @@
+
+{-# OPTIONS_GHC -Wno-missing-export-lists #-}
+
+module Database.LMDB.FFI where
```
Add a note why I'm exporting everything

jorisdral replied:
Added an export list instead

jorisdral commented on cardano-lmdb.cabal:
```diff
@@ -20,26 +20,43 @@ license:       BSD-2-Clause
+    Database.LMDB.FFI
+    Database.LMDB.Raw
+
+  pkgconfig-depends:  lmdb >=0.9 && <0.10s
```
```suggestion
  pkgconfig-depends:  lmdb >=0.9 && <0.10
```

jasagredo replied:
```suggestion
  pkgconfig-depends:  lmdb >=0.9 && <0.10
```
?

jasagredo commented on .github/workflows/haskell.yml:
```diff
@@ -4,21 +4,36 @@ on:
-        ghc: ["8.10.7", "9.2.5"]
-        cabal: ["3.8.1.0"]
+        ghc: ["8.10.7", "9.2.8", "9.4.8", "9.6.4", "9.8.2"]
+        cabal: ["3.10.2.1"]
```
We can jump to `3.10.3.0`

jasagredo commented on .github/workflows/haskell.yml:
```diff
@@ -4,21 +4,36 @@ on:
+        ghc: ["8.10.7", "9.2.8", "9.4.8", "9.6.4", "9.8.2"]
+        cabal: ["3.10.2.1"]
         os: [ubuntu-latest, windows-latest, macOS-latest]
+        exclude:
+          - ghc: "9.4.8"
+            os: windows-latest
+          - ghc: "9.4.8"
+            os: macOS-latest
+          - ghc: "9.8.2"
+            os: windows-latest
+          - ghc: "9.8.2"
+            os: macOS-latest
```
Why?

jorisdral replied:
Removed

jasagredo commented on cardano-lmdb.cabal:
```diff
@@ -20,26 +20,43 @@ license:       BSD-2-Clause
+  import:             warnings
+  hs-source-dirs:     hsrc_lib
+  default-language:   Haskell2010
+  build-depends:      base >=4.14 && <20
```
```suggestion
  build-depends:      base >=4.14 && <4.20
```

jasagredo commented on cardano-lmdb.cabal:
```diff
@@ -20,26 +20,43 @@ license:       BSD-2-Clause
-                        -Wmissing-export-lists
-                        -Wno-unticked-promoted-constructors
+  import:             warnings
+  hs-source-dirs:     hsrc_lib
```
Could also use `src` which is usually the name

jasagredo commented on hsrc_lib/Database/LMDB/FFI.hsc:
```diff
@@ -0,0 +1,293 @@
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+{-# OPTIONS_GHC -Wno-missing-export-lists #-}
```
I think you could instead do `module Database.LMDB.FFI (..) where`, no?

jorisdral replied:
That didn't seem to work -- instead, I added a full export list

jasagredo commented on hsrc_lib/Database/LMDB/FFI.hsc:
```diff
@@ -0,0 +1,293 @@
+foreign import capi "lmdb.h mdb_env_close" mdb_env_close :: Ptr MDB_env -> IO ()
+foreign import capi "lmdb.h mdb_env_set_flags" mdb_env_set_flags :: Ptr MDB_env -> CUInt -> CInt -> IO CInt
+foreign import capi unsafe "lmdb.h mdb_env_get_flags" mdb_env_get_flags :: Ptr MDB_env -> Ptr CUInt -> IO CInt
+foreign import ccall unsafe "lmdb.h mdb_env_get_path" mdb_env_get_path :: Ptr MDB_env -> Ptr (Ptr CChar) -> IO CInt
```
Perhaps use CPP to define also the `capi` version? for removing the `ccall` one eventually

jorisdral replied:
I've tried that, but then you get cascading changes, because the type of the function changes depending on the CPP branch and there are also conditional changes on imports. I think it's fine to keep this ccall for now, since we don't really this function anyway

jasagredo commented on hsrc_lib/Database/LMDB/Raw.hs:
```diff
@@ -1,4 +1,9 @@
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{- HLINT ignore "Use camelCase" -}
```
Probably define in a `.hlint.yaml` file? it looks reasonable if dealing with FFI

jorisdral replied:
It's only for this file, so I'll probably leave it as is

jasagredo commented on .github/workflows/haskell.yml:
```diff
@@ -4,20 +4,26 @@ on:
-        cabal: ["3.8.1.0"]
+        ghc: ["8.10.7", "9.2.8", "9.4.8", "9.6.4", "9.8.2"]
+        cabal: ["3.10.3.0"]
         os: [ubuntu-latest, windows-latest, macOS-latest]
```
I'm now thinking that we might actually want to use the `sys {os, shell}` approach that was there in lmdb-simple already, i.e. default the Windows shell to the msys2 shell already

jorisdral replied:
Changed in 068d068c9632295e66512b2176fa54c2022455d0. Can I squash and merge like this?

jorisdral committed:
Refactor FFI internals

* General cleanup
* Switch to capi calling convention
* Correct and simplify C-types
* Move FFI code to a separate `FFI` module in a private sub-library
* Add version macros, environment flags, database
  flags, write flags, return codes, and cursor get
  operations as `CapiFFI` values to the FFI module.
* Remove rogue occurrences of `#const` and
  `#type` in the `Raw` module.
* Remove array dependency
* Add simple unit tests for low-level and higher-level bindings

jorisdral committed:
Clean up unused files, update documentation

jorisdral committed:
Update GHA workflow, add an index-state to cabal.project
